### ISSUE TITLE
how to support Luajit

### ISSUE THREAD

**Issue by heijiu** [h2o/h2o#616]
how to support Luajit like Nginx-module-lua

**Comment by kazuho** [comment]
FWIW, H2O already has [support for mruby](https://h2o.examp1e.net/configure/mruby_directives.html), and regarding your question, it would be worthwhile to look into how it is implemented in [lib/handler/mruby.c](https://github.com/h2o/h2o/blob/master/lib/handler/mruby.c).

**Comment by nikolay** [comment]
@kazuho That's great, but isn't mruby times slower than LuaJIT? Unlike mruby, which is a subset of Ruby, LuaJIT is full Lua implementation and it's widely used in Nginx at high-volume service providers such as CloudFlare, for example. I think if you guys add first-class LuaJIT support, H2O can then truly replace Nginx in many cases!

**Comment by kazuho** [comment]
@nikolay Thank you for the suggestion!

I am more than happy to support others trying to implement lua handler on H2O.  And for such purpose, my suggestion is to mimic the design and implementation of the mruby handler, since the two will have many aspects in common.

OTOH, FWIW a JIT exists for mruby by @miura1729 (https://github.com/miura1729/mruby), though AFAIK it unfortunately only supports 32-bit x86 systems; also mruby is not a subset but in fact a full implementation of the ruby programming language (ISO/IEC 30170:2012) with some additions.

**Comment by nikolay** [comment]
@kazuho Thanks for the clarification. As a "lightweight Ruby", I thought mruby lacked language features.

**Comment by mingodad** [comment]
Hello !
I'm trying to add lua support to h2o-cpp and I'm getting this error:
```
lib/common/socket/evloop.c.h:269: void do_write(h2o_socket_t*, h2o_iovec_t*, size_t, h2o_socket_cb): Assertion `sock->super._cb.write == __null' failed.
received fatal signal 6;
```
It happens after make a request to this simple hello.lua
```
--per request call
function h2oManageRequest(req)
	local host = req:host() --also req:authority()
	local path = req:path()
	local name = path --path:sub(after_lua_prefix_len)
	req:send("Hello " .. name, "text/plain; charset=utf-8")
	return 0
end
```
The code behind it is similar to a simple echo module I did and works fine:
```
static int lua_h2o_req_send(lua_State *L)
{
    CHECK_H2O_REQUEST();
    static h2o_generator_t generator = { NULL, NULL };

    h2o_iovec_t body;
    body.base = (char*)luaL_checklstring(L, 2, &body.len);

    int is_final = 1;

    if(lua_isstring(L, 3))
    {
        h2o_iovec_t content_type;
        content_type.base = (char*)lua_tolstring(L, 3, &content_type.len);
        req->res.headers.add(&req->pool, H2O_TOKEN_CONTENT_TYPE, content_type.base, content_type.len);
        req->res.content_length = body.len;
        req->res.status = 200;
        req->res.reason = "OK";
    } else {
        is_final = luaL_optint(L, 3, 1);
    }
    if (!req->_generator) {
        req->start_response(&generator);
    }
    req->send(&body, 1, is_final);
    return 0;
}
```
The browser receives the request and it seems that the request somehow is processed again and fail.
I know that this is a fork and doesn't belong to h2o but if the kazuho could give some light on this error (maybe explaining the workflow of a request processing in a document, it would be nice if the workflow for a module also could be explained on a document).

Cheers !

**Comment by mingodad** [comment]
Without any kind of documentation of how h2o is intended to work only reading the code and guessing would make it hard for third parties write anything for h2o.

**Comment by kazuho** [comment]
@mingodad Glancing at your code, it might be due to the fact that you are not guarding the strings obtained from lua from getting GCed.

Guarding the values (or duplicating them outside of lua) is necessary, since `lua_tolstring` (and `lua_checklstring`) do not guarantee _that the pointer returned by lua_tolstring will be valid after the corresponding value is removed from the stack_ (see http://pgl.yoyo.org/luai/i/lua_tolstring) and that because H2O is an event-driven server, the protocol handlers of H2O may send the headers and body after the call to the handler function returns.

It should also be noted that unless you are sending all contents at once (by setting `is_final` to 1), you need to create your own generator object and do the synthesize the call to h2o_send with the core calling the `proceed` callback.  See #310.

Anyways, I strongly suggest to first create a stable handler that does the minimum (i.e. call a script that returns a response and send the response all at once using `h2o_send_inline`) as was done in the [mruby handler of version 1.6.0](https://github.com/h2o/h2o/blob/v1.6.0/lib/handler/mruby.c#L517-L549), and then try to add more features step by step.

If you try to do many things at once, it becomes very hard to determine what is wrong.

**Comment by mingodad** [comment]
Hello !
Thanks for replying !
I just now found the problem !
I was due to "return 0" from lua but testing it as boolean in C (0 is not false).
The strings been garbage collected is another problem too, thanks !
Cheers !

**Comment by frink** [comment]
I'm wondering whether scripting support externally might be a better approach to internally. Essentially, creating a Node type framework where the config file is the scripting language. This is what Lua was designed for in the first place as a quasi-scripting/config language. Seems like creating a wrapper that would call h2o where necessary via ffi might be more flexible than trying to pidgin-hole Lua and take away a lot of its power...

Might also be much simpler to write a FFI for libh2o instead of trying to embed Lua.

Dunno. Just a thought...

**Comment by frink** [comment]
Originally Lua was just a scripting language for the Brazilian oil company. But it's interop and small footprint make it special for so many fun things. Wow configuration not the least of them. Esp. since the Lua interpreter is smaller than most YAML libraries...

BTW: LuaJIT works fine for me on MacOS. But the compilation of LuaJIT on MacOS is truly a pain. Actually, I hate compiling anything on MacOS because their xcode junk always screws something up...

**You're right that people don't want Lua + H2O for configuration per se...**
Lua is a great language for web server scripting because of its FFI - although within a multithreaded/fiber/coroutine environment Lua gives you more rope to hang yourself than most other languages. Nginx, Ardour, and Postgres documentation all talk a lot about this in their implementations.

This was why I suggested the FFI wrapper instead of plugging Lua inside H2o's context. The FFI approach allows Lua to not be pigeon holed into a lot of concepts that are very un-Lua-like which cause buggy implementations. Nginx particularly, has some weird problems with passing data between execution stages that make Lua apps very hard to debug. Don't do things like Nginx or you will handicap Lua in a huge way.

Lua is awesome for writing small apps. Smokes every other non-compiled language as far as security and speed. Great for writing microservices. If you're going to implement Lua inside of H2O you will need to override it's coroutine() structures so that it allows H2O to manage the execution properly. That can be much easier said than done. But several people have done a decent job with overriding the native coroutines...

All that to say: Good luck!

**Comment by metaminimalist** [comment]
[luajit is also unfortunately not actively developed and maintained any more,](https://www.freelists.org/post/luajit/Looking-for-new-LuaJIT-maintainers) the project owner more or less stepped back and [even rejects sponsoring of new features.](http://luajit.org/sponsors.html)

which means that luajit did not follow lua sind 4-5 years
which means that it suffers /will suffer from bitrot

why would you want to have it in your webserver?

**Comment by Qix-** [comment]
@metaminimalist Because it's very fast compared to PUC's and adheres to Lua 5.1 which is widely considered a stable version of the language.

**Comment by frink** [comment]
@metaminimalist it's still maintained. You've been misinformed... See: https://github.com/LuaJIT/LuaJIT/commits/master

Please note LuaJIT 5.1 is what most web tools are using are embedding. A large part of CloudFlare and Google infrastructures are built on LuaJIT so it's hardly obsolete in any sense of the word.

**Comment by Qix-** [comment]
Yeah the two links mentioned are only the original author stepping down. Maintainers step down all the time, and is normally seen as a good thing since there's usually a good reason.

**Comment by frink** [comment]
When you dig deeper into the language you understand why LuaJIT has not been quick to import the 5.2 language "improvements"... Some of them cause huge challenges for the clean ABI that LuaJIT is famous for...

**Comment by Qix-** [comment]
@frink I've always wondered, but that makes sense. Could you list a few of them?

**Comment by frink** [comment]
Gosh it's been a while. Can't think of the top of my head. But I remember reading some pretty heavy discussions about it a few years ago. I'm sure the articles are still out there but I can't remember off the top of my head... Sorry.

**Comment by Qix-** [comment]
> Mike's expressed displeasure with some of them (e.g. the details of
> the 5.2 bit library), and approved of others (e.g., "goto").
> I suspect luajit will likely over time add various 5.2 features as
> time and funding allow (and as clients request, of course).

Interesting, thanks for the info either way. Gives me something to research :)

If anyone is interested, I've started a luajit implementation here: https://github.com/Qix-/h2o/tree/luajit

**Comment by frink** [comment]
There are some things in the 5.2 security model that are a huge change. Deep internals that most won't use. But when you're trying to extend Lua as a framework inside of Lua there are some huge differences. I just don't remember if those are the differences that Paul had issue with or not.

Generally, I think I remember something about state needing to be maintained in weird ways with the new model. I know Lua tried to go more OO in 5.2 and I think some of that may have caused issue with the JIT interpreter thus it was refused. It was something to that effect. Paul had some good points.

LuaJIT today should be considered as a cousin of Lua and not exactly the same thing...
